src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java

```diff
@@ -1,60 +1,152 @@
 package com.bobocode.hoverla.bring.context;
 
+import com.bobocode.hoverla.bring.exception.NoSuchBeanException;
+import com.bobocode.hoverla.bring.exception.NoUniqueBeanException;
+import com.google.common.collect.HashBasedTable;
+import com.google.common.collect.Table;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.Objects;
+
 
 /**
- * Example of ApplicationContext implementation, can be removed or reimplemented when actual development starts
+ * Implementation of ApplicationContext
  */
+@Slf4j
 public class ApplicationContextImpl implements ApplicationContext {
+    private Table<String, Class<?>, BeanDefinition> context = HashBasedTable.create();
+    private final String BEAN_TYPE_MUST_BE_NOT_NULL_MESSAGE = "The argument [beanType] must be not null";
+    private final String NO_SUCH_BEAN_EXCEPTION_MESSAGE = "Bean with provided name/type [%s] not found in the context";
+
+    /**
+     *  Scanners scan application to define beans configs and map them into {@link BeanDefinition} lists.
+     *  Received list of bean definitions validates by validator to avoid duplicates and other problems (see {@link BeanValidator}).
+     *  Beans definitions list maps to container structure. It's Table<String, Class<?>, BeanDefinition>, where
+     *      - row key is a name of bean
+     *      - column key is a type of bean
+     *      - mapped value is bean definition
+     *  Then context passes to {@link BeanInitializer} for beans instances initialization
+     *
+     * @param scanners      list of scanners for
+     * @param validator     bean definition validator
+     * @param initializer   bean initializer
+     */
     public ApplicationContextImpl(List<BeanScanner> scanners,
                                   BeanValidator validator,
                                   BeanInitializer initializer) {
-        /* 1) Scanners get all beans
-           2) Merge List of bean definitions into one List
-           3) Send beans list to validator to check duplicates
-           4) Convert bean list to map
-           5) Send beans map to initializer to instantiate beans and inject dependencies
-         */
+
+        List<BeanDefinition> beanDefinitionList = scanningApplication(scanners);
```

Comment:
1. Usually method name should not be a noun. In most cases it should be a verb (e.g. `scanApplication()`)
2. Regarding method names - they should tell the reader what exactly method does - `scanApplication()` in this case will not work, since we're no scanning any application. We're just triggering `scan()` method in every `BeanScanner`. Thus I would name this method smth like `runScanners` (feel free to come up with a better name).
3. Now I would like to outline, that methods `validate()` and `beanInstantiation()` do nothing except for logging some message and delegate method call. Therefore, I would remove these since you can call them directly in constructor.
4. Method `contextCreation()` is also not correctly named, since you are just populating the container table with bean definitions. I would remove direct table instantiation from line 22 and change line 45 to `this.beanDefinitionTable = populateBeanDefinitionTable(beanDefinitionList)`. After that you can change line 46 to `initializer.initialize(beanDefinitionTable)`.

Having said that, I would change the constructor to look like this: 
```
List<BeanDefinition> beanDefinitionList = runScanners(scanners);
validator.validate(beanDefinitionList);
this.beanDefinitionTable = populateBeanDefinitionTable(beanDefinitionList);
initializer.initialize(beanDefinitionTable);
```
Feel free to contact me for discussion ;)

Reply:
Thank you for the detailed explanation. Honestly, this piece of code doesn't look good to me neither. But I didn't know what to change.

Reply:
Done